accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,126 @@
+
+The .NET SDK already does conflict resolution for other asset types: references and copy-local runtime assets.  It should do the same for Analyzers.  Analyzers need not have "platform manifest" inputs for conflict resolution since all analyzers will be available during build, unlike runtime files which may not be available.  Comparison for the sake of conflict resolution can consider the same assembly and file version rules as reference files.  When finding two or more source generators with the same name
+
+Source generator conflict resolution should run before compile, but after all NuGet package assets have been evaluated.  Specific sequencing, public/private target naming, extensibility will not be specified here but should follow best practice as determined by the SDK team. 
+
+### Source generator versioning
+
+Source generators should be strongly named and update assembly version with every public stable release (GA and servicing).  This ensures that the publicly shipped source generators can be loaded side-by-side in the same process and garuntees 
+
+### Source generator compatibility concerns
+
+Source generators should only use public API in framework components they extend.  This public API already needs to meet the strict binary compatibility guarantees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.
+
+Source generators should not use public API with "reserved" or "undocumented" functionality in attempts to avoid public API guarantees.  Libraries paired with source generators need to provide the same guarantees as libraries alone do today.
+
+Source generators should avoid emitting public API in user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.
+
+Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name colisions with user code and provide affordances for resolving name conflicts. 
```

Comment:
Is there any general guidance on how the generated code should look like (e.g. name prefix, special attributes, version suffix, etc)?

Reply:
Have you thought about using other version specific features such as nullability? Razor side-steps documentation and  nullability by pragma suppressing them.

Reply:
There might be a debuggability connection here, eg., if the code is marked `<autogenerated />`

In general I have not seen what the debuggability story is for source generators. If you are debugging on the same machine you generated on, I guess the PDB points to a generated file in temp (which I believe is generated only for this purpose). But what is the story if you are debugging on a different machine? Is the source serialized into the PDB? I have not seen details of this, perhaps you have a link.

Reply:
cc @cartermp

Reply:
I don't intend to go into super detail around strategies here.  If possible we should like to design considerations the compiler takes when doing similar things.  I guess the unique problem here is that the compiler usually generates code that it doesn't expect callers to interact with vs source generators specifically generate code that the user needs to interact with.

Reply:
> Is the source serialized into the PDB?

Yes, it should be.  You can also inspect it in the IDE, or even have the compiler write to disk with `EmitCompilerGeneratedFiles`.  I couldn't find the docs on this, @chsienki or @jasonmalinowski might know.

> Have you thought about using other version specific features such as nullability? Razor side-steps documentation and nullability by pragma suppressing them.

We should annotate the generated code so that it works well with user's types, be they nullable or not.  I will add a note.  In general we should be trying to write the *best code possible* rather than coming up with reasons for not following the rules, except maybe for tabs vs spaces.  That rule should of course be suppressed ;)